### ISSUE TITLE
DateOffset around end of February does not reproduce consecutive dates

### ISSUE THREAD

**Issue by pjarnhus** [pandas-dev/pandas#22929]
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame([1,2,3], index=pd.DatetimeIndex(start='2018-02-27', periods=3, freq='D'), columns=['A'])
df.index
```
returns
`DatetimeIndex(['2018-02-27', '2018-02-28', '2018-03-01'], dtype='datetime64[ns]', freq='D')`

Performing the following offset
```python
df.index = df.index + pd.tseries.offsets.DateOffset(months=4, days=2)
df.index
```
returns
`DatetimeIndex(['2018-06-29', '2018-06-30', '2018-07-03'], dtype='datetime64[ns]', freq=None)`
#### Problem description
My apologies if this has already been posted, I was not able to find a similar issue among the ones posted.

When I, rather naively, tried to offset a DatetimeIndex, I ended up with a few missing dates. This could be traced back to the jump from Feb 28 to Mar 1. I would have expected that applying an offset to the DatetimeIndex object would offset every element in the index according to the needed offset of the first entry. Instead the offset seems to be applied to every element individually

The only way I could achieve this was by doing
```python
df.index = df.tshift((df.index[0]+pd.tseries.offsets.DateOffset(months=4,days=2)-df.index[0]).days).index
```
which seems rather clunky.

#### Expected Output
`DatetimeIndex(['2018-06-29', '2018-06-30', '2018-07-01'], dtype='datetime64[ns]', freq=None)`
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.6.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 61 Stepping 4, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.4
pytest: 3.7.3
pip: 18.0
setuptools: 40.2.0
Cython: 0.28.5
numpy: 1.15.1
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.5.0
sphinx: 1.7.7
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 2.2.3
openpyxl: 2.5.5
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.8
lxml: 4.2.4
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.11
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

**Comment by mroeschke** [comment]
This looks like the expected behavior.

DatetimeIndexes are not assumed to always be monotonic, so offsetting relative to the first entry would produce odd results. Also your last, expected date would only be shifted by 4 months ignoring the 2 days.

**Comment by pjarnhus** [comment]
Fair enough. It could be nice to have a comment in the documentation regarding this assumption then. 

Thank you for the swift reply.

**Comment by mroeschke** [comment]
Documentation regarding DateOffset and months would be very welcome.

**Comment by jbrockmendel** [comment]
Might make sense to let DatetimeIndex.shift take a DateOffset, shifting the first date by the given DateOffset and then retaining the original freq.
